File: homeworks/homework3/src/count_sort.py
#! /usr/bin/env python

from __future__ import print_function, division

def count_sort(a):
    temp = a[:]

    for i, ai in enumerate(a):
        count = 0
        for j, aj in enumerate(a):
            if aj < ai:
                count += 1
            elif (aj == ai) and (j < i):
                count += 1
        temp[count] = ai

    return temp


if __name__ == "__main__":
    tests = [
      [0,1,2,3],
      [3,2,1,0],
      [3,1,2,0],
    ]

    for a in tests:
        a_sort = count_sort(a)
        print(a_sort == list(sorted(a)))

File: homeworks/homework3/src/count_sort.c
void count_sort(int a[], int n) {
  int i, j, count;
  int *temp = malloc(n*sizeof(int));
  for (i = 0; i < n; i++) {
    count = 0;
    for (j = 0; j < n; j++) {
      if (a[j] < a[i])
        count++;
      else if (a[j] == a[i] && j < i)
        count++;
    }
    temp[count] = a[i];
  }
  memcpy(a, temp, n*sizeof(int));
  free(temp);
} /* Count sort */


int main(int argc, char** argv) {
  return 0;
}
